File: apps/android/app/src/test/java/com/elite/clinic/sync/SessionFrameCodecTest.kt
package com.elite.clinic.sync

import android.util.Base64
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class SessionFrameCodecTest {
    private val sessionId = "session-frame-01"
    private val noncePrefix = hex("01020304")
    private val clientKey = ByteArray(32) { 0x11 }
    private val hubKey = ByteArray(32) { 0x22 }

    @Test
    fun encryptsAndDecryptsAesGcmFrameWithCounterNonce() {
        val client = clientCodec()
        val hub = hubCodec()
        val frame = client.encrypt(
            "sync-request",
            "{\"scope\":\"appointments\"}".toByteArray(),
        )
        assertEquals(
            Base64.encodeToString(hex("010203040000000000000000"), Base64.NO_WRAP),
            frame.getString("nonceBase64"),
        )
        assertEquals(
            "{\"scope\":\"appointments\"}",
            String(hub.decrypt(frame)),
        )
        assertEquals(1L, hub.nextReceiveCounter())
    }

    @Test
    fun rejectsReplayCounterGapNonceAndAuthenticationTampering() {
        val client = clientCodec()
        val hub = hubCodec()
        val first = client.encrypt("sync-request", "first-session-payload".toByteArray())
        expectFailure("ELITE_SESSION_COUNTER_GAP") {
            hub.decrypt(JSONObject(first.toString()).put("counter", 1L))
        }
        assertEquals("first-session-payload", String(hub.decrypt(first)))
        expectFailure("ELITE_SESSION_REPLAY_REJECTED") { hub.decrypt(first) }

        val second = client.encrypt("sync-request", "second-session-payload".toByteArray())
        expectFailure("ELITE_SESSION_NONCE_MISMATCH") {
            hub.decrypt(JSONObject(second.toString()).put("nonceBase64", "AAAAAAAAAAAAAAAA"))
        }
        expectFailure("ELITE_SESSION_AUTHENTICATION_FAILED") {
            hub.decrypt(
                JSONObject(second.toString()).put(
                    "ciphertextBase64",
                    Base64.encodeToString("tampered-ciphertext".toByteArray(), Base64.NO_WRAP),
                ),
            )
        }
    }

    @Test
    fun closesAndRejectsFurtherUse() {
        val client = clientCodec()
        client.close()
        try {
            client.encrypt("sync-request", "closed".toByteArray())
            throw AssertionError("expected ELITE_SESSION_CLOSED")
        } catch (error: SecurityException) {
            assertTrue(error.message?.contains("ELITE_SESSION_CLOSED") == true)
        }
    }

    @Test
    fun derivesAndVerifiesKeyConfirmationMac() {
        val keys = SessionKeyDerivation.deriveSessionKeys(ByteArray(32) { 3 }, ByteArray(32) { 7 })
        val mac = SessionKeyDerivation.keyConfirmationMac(
            keys.keyConfirmationKey,
            "session-frame-01",
            "a".repeat(64),
            "client",
        )
        assertTrue(
            SessionKeyDerivation.verifyKeyConfirmation(
                mac,
                SessionKeyDerivation.keyConfirmationMac(
                    keys.keyConfirmationKey,
                    "session-frame-01",
                    "a".repeat(64),
                    "client",
                ),
            ),
        )
    }

    private fun clientCodec() = SessionFrameCodec(
        sessionId = sessionId,
        noncePrefix = noncePrefix,
        sendKey = clientKey.copyOf(),
        receiveKey = hubKey.copyOf(),
        sendDirection = "client-to-hub",
        receiveDirection = "hub-to-client",
    )

    private fun hubCodec() = SessionFrameCodec(
        sessionId = sessionId,
        noncePrefix = noncePrefix,
        sendKey = hubKey.copyOf(),
        receiveKey = clientKey.copyOf(),
        sendDirection = "hub-to-client",
        receiveDirection = "client-to-hub",
    )

    private fun expectFailure(code: String, block: () -> Unit) {
        try {
            block()
            throw AssertionError("expected $code")
        } catch (error: IllegalArgumentException) {
            assertTrue(error.message?.contains(code) == true)
        }
    }

    private fun hex(value: String): ByteArray = value.chunked(2)
        .map { it.toInt(16).toByte() }
        .toByteArray()
}

File: packages/auth/src/session-frame-codec.test.ts
import { describe, expect, it } from "vitest";
import {
  SessionFrameChannel,
  deriveSessionFrameNonce,
} from "./session-frame-codec.js";

const sessionId = "session-frame-01";
const noncePrefix = Buffer.from("01020304", "hex");
const clientKey = Buffer.alloc(32, 0x11);
const hubKey = Buffer.alloc(32, 0x22);

function channels() {
  return {
    client: new SessionFrameChannel({
      sessionId,
      noncePrefix,
      sendKey: clientKey,
      receiveKey: hubKey,
      sendDirection: "client-to-hub",
      receiveDirection: "hub-to-client",
    }),
    hub: new SessionFrameChannel({
      sessionId,
      noncePrefix,
      sendKey: hubKey,
      receiveKey: clientKey,
      sendDirection: "hub-to-client",
      receiveDirection: "client-to-hub",
    }),
  };
}

describe("SessionFrameChannel", () => {
  it("encrypts and decrypts AES-GCM frames with deterministic counter nonces", () => {
    const { client, hub } = channels();
    const frame = client.encrypt(
      "sync-request",
      Buffer.from('{"scope":"appointments"}', "utf8"),
    );
    expect(frame.counter).toBe(0);
    expect(frame.nonceBase64).toBe(
      Buffer.from("010203040000000000000000", "hex").toString("base64"),
    );
    const decrypted = hub.decrypt(frame);
    expect(decrypted.plaintext.toString("utf8")).toBe(
      '{"scope":"appointments"}',
    );
    expect(hub.nextReceiveCounter).toBe(1);
  });

  it("rejects replay, gaps, nonce tampering, and ciphertext authentication failures", () => {
    const { client, hub } = channels();
    const first = client.encrypt(
      "sync-request",
      Buffer.from("first-session-payload"),
    );
    expect(() => hub.decrypt({ ...first, counter: 1 })).toThrow(
      "ELITE_SESSION_COUNTER_GAP",
    );
    expect(hub.decrypt(first).plaintext.toString()).toBe(
      "first-session-payload",
    );
    expect(() => hub.decrypt(first)).toThrow("ELITE_SESSION_REPLAY_REJECTED");

    const second = client.encrypt(
      "sync-request",
      Buffer.from("second-session-payload"),
    );
    expect(() =>
      hub.decrypt({ ...second, nonceBase64: "AAAAAAAAAAAAAAAA" }),
    ).toThrow("ELITE_SESSION_NONCE_MISMATCH");
    expect(() =>
      hub.decrypt({
        ...second,
        ciphertextBase64: Buffer.from("tampered-ciphertext").toString("base64"),
      }),
    ).toThrow("ELITE_SESSION_AUTHENTICATION_FAILED");
  });

  it("rejects invalid nonce prefixes and unsafe counters", () => {
    expect(() => deriveSessionFrameNonce(Buffer.alloc(3), 0)).toThrow(
      "ELITE_SESSION_NONCE_PREFIX_INVALID",
    );
    expect(() =>
      deriveSessionFrameNonce(noncePrefix, Number.MAX_SAFE_INTEGER + 1),
    ).toThrow("ELITE_SESSION_COUNTER_INVALID");
  });
});

import { readFileSync } from "node:fs";
import { resolve } from "node:path";

describe("SessionFrameChannel shared vectors", () => {
  it("matches the repository AES-GCM vector exactly", () => {
    const fixture = JSON.parse(
      readFileSync(
        resolve(process.cwd(), "../../test-vectors/session-frame-vectors.json"),
        "utf8",
      ),
    ) as {
      vectors: Array<{
        keyHex: string;
        noncePrefixHex: string;
        plaintextBase64: string;
        frame: Parameters<SessionFrameChannel["decrypt"]>[0];
      }>;
    };
    const vector = fixture.vectors[0]!;
    const channel = new SessionFrameChannel({
      sessionId: vector.frame.sessionId,
      noncePrefix: Buffer.from(vector.noncePrefixHex, "hex"),
      sendKey: Buffer.from(vector.keyHex, "hex"),
      receiveKey: Buffer.alloc(32, 0x22),
      sendDirection: "client-to-hub",
      receiveDirection: "hub-to-client",
    });
    const actual = channel.encrypt(
      vector.frame.messageType,
      Buffer.from(vector.plaintextBase64, "base64"),
    );
    expect(actual).toEqual(vector.frame);
  });
});
